man/sdhash-cli.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SDHASH-CLI 1"
.TH SDHASH-CLI 1 "2013-07-15" "" ""
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
sdhash\-cli \- client for similarity digest server
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBsdhash-cli\fR [options] [mode] [files]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBsdhash-cli\fR is a client front end to the sdhash server sdhash-srv.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-H\fR, \fB\-\-host\fR <hostname>" 4
.IX Item "-H, --host <hostname>"
Connects to sdhash server running at host.  Client defaults to localhost.
.IP "\fB\-P\fR, \fB\-\-port\fR <port>" 4
.IX Item "-P, --port <port>"
Sets port of sdhash server to connect to.  Client defaults to port 9090.
.IP "\fB\-b\fR, \fB\-\-block\-size\fR <N>" 4
.IX Item "-b, --block-size <N>"
Enables block mode for hash generation.  Hashes will be generated in Nkb blocks.
Suggested block size for disk hashes is 16, for memory dumps we suggest 4.
.IP "\fB\-t\fR, \fB\-\-threshold\fR <\-1\-100>" 4
.IX Item "-t, --threshold <-1-100>"
Sets a confidence threshold for comparisions.  Use this to filter results
to reduce false positives.  Used with comparison and query modes.
Use '\-1' to display all comparisons.
.IP "\fB\-s\fR, \fB\-\-sample\-size\fR <1\-16>" 4
.IX Item "-s, --sample-size <1-16>"
Uses a sampling of filters to perform queries, which speeds up
query results significantly.  Used for query-mode only.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Displays warnings
.SH "MODES"
.IX Header "MODES"
.IP "\fB\s-1HASHING\s0,LOADING,SAVING,LISTING\fR" 4
.IX Item "HASHING,LOADING,SAVING,LISTING"
.PD 0
.IP "\fB\-l\fR, \fB\-\-list\fR" 4
.IX Item "-l, --list"
.PD
List sets of hashes currently loaded on the running server.
.IP "\fBdefault\fR <files>" 4
.IX Item "default <files>"
The default mode is generation mode, and will generate similiarity digests of 
all files given to it on the command line and save them to a set.
Using \-b will generate block-based hash digests which are useful for processing disk images and memory dumps.
.RS 4
.Sp
.RS 4
\&\fBsdhash-cli\fR [options] file1 file2 file4 foo*
.Sp
\&\fBsdhash-cli\fR \fB\-\-block\-size\fR 16 image.dd
.RE
.RE
.RS 4
.RE
.IP "\fB\-i\fR, \fB\-\-import\fR <file.sdbf>" 4
.IX Item "-i, --import <file.sdbf>"
Imports a sdbf file into the server.
.IP "\fB\-e\fR, \fB\-\-export\fR <N> <file.sdbf>" 4
.IX Item "-e, --export <N> <file.sdbf>"
Exports set number N to file.sdbf.
.IP "\fB\-d\fR, \fB\-\-show\-set\-data\fR <N>" 4
.IX Item "-d, --show-set-data <N>"
Prints all SDBFs in set number N to stdout.
.IP "\fB\-n\fR, \fB\-\-show\-set\-names\fR <N>" 4
.IX Item "-n, --show-set-names <N>"
Prints names and sizes of all SDBFs in set number N to stdout.
.IP "\fB\s-1COMPARING\s0\fR" 4
.IX Item "COMPARING"
.PD 0
.IP "\fB\-c\fR, \fB\-\-compare\fR <N>" 4
.IX Item "-c, --compare <N>"
.PD
Single comparison mode takes a set number as an argument and compares each item in that set to every other item in that set.
.RS 4
.Sp
.RS 4
\&\fBsdhash-cli\fR \fB\-\-compare\fR 3
\&\fBsdhash-cli\fR \fB\-c\fR 3
.RE
.RE
.RS 4
.RE
.IP "\fB\-c\fR, \fB\-\-compare\fR <N> <M>" 4
.IX Item "-c, --compare <N> <M>"
Comparison mode can also compare two sets, numbers N and M, optionally sampling the \*(L"query\*(R" with the \fB\-s\fR sample size
option, and compares each digest in the query to each digest in the target.
.RS 4
.Sp
.RS 4
\&\fBsdhash-cli\fR \fB\-\-compare\fR 3 5
.Sp
\&\fBsdhash-cli\fR \fB\-\-sample\-size\fR 4 \fB\-\-compare\fR 0 4
.Sp
\&\fBsdhash-cli\fR \fB\-s\fR 4 \fB\-c\fR 2 7
.RE
.RE
.RS 4
.RE
.IP "\fB\-\-results\-list\fR" 4
.IX Item "--results-list"
Retrieve a listing of results
.IP "\fB\-\-type\fR <default|indexing|web>" 4
.IX Item "--type <default|indexing|web>"
Sets the type of results listing searched for.  If omitted, will search
for comparison default results.
.IP "\fB\-r\fR, \fB\-\-results\fR resultID" 4
.IX Item "-r, --results resultID"
Returns saved results, if any, associated with a resultID retrieved
from \fB\-\-results\-list\fR
.IP "\fB\s-1SEARCHING\s0\fR" 4
.IX Item "SEARCHING"
Searching uses index files generated upon hashing or loaded into the
server to provide a fast lookup mechanism useful for driving data filtering.
Searching does not generate a saved set of digests; it generates a results set.
Results can be retrieved by using \fB\-\-results\-list\fR \fB\-\-type\fR indexing.
.IP "\fB\-\-index\-search\fR" 4
.IX Item "--index-search"
Produces matches by searching set-level indexes.
.IP "\fB\-\-index\fR" 4
.IX Item "--index"
Hashing option, generate and save indexes while creating sdbf hashes.
.SH "AUTHORS"
.IX Header "AUTHORS"
\&\fBsdhash-cli\fR was written by Vassil Roussev <vassil@roussev.net> and Candice Quates <candice@egobsd.org>.  http://sdhash.org/
.SH "LICENSE"
.IX Header "LICENSE"
\&\fBsdhash-cli\fR is distributed under the Apache License v2.0


ChangeLog
New in version 3.4

New constructor for sdbf class to accept strings. 

Major GPU program speedup via reduction changes.

Bugfix for current boost version and c++ 11

Bugfix for underflow in hash generation 

Updated/tested python swig interface

x86-32 bit linux compilation fix for cpuid

Updated test program.

Thanks to Tom Sires, Alex Nelson, Jesse Kornblum, Simon Spero, and Andrei Costin for
patches and requests.

--------
New in version 3.3

Major bugfix involving self-to-self block-based comparison scoring,
we are now evenly discounting unfilled filters.

Added --separator change option.

Clarified indexing options/formats.

Enabled support for multiple lists-of-files to hash.

Now passing -Wall on gcc on Linux in main program.

--------
New in version 3.2.5

Significant speedup of CPU comparisons via openMP threads.  Now 
requiring openMP. 

Program will automatically use the maximum number of cores/threads
as given by openMP, on generation and comparison.
--------
New in version 3.2

CUDA-accelerated GPU comparison program available - sdhash-gpu. 

Changes to indexing to auto-generate indexes of larger, more 
usable size (640MB) to search much more information.

Small change to bloom filter fullness threshold for comparison,
we now require 16 elements instead of 6.
--------
New in version 3.1

New comparison code using the POPCNT instruction.  Marked speed improvements 
on intel, cache performance improvements on AMD.

Refined the searching/index components, including auto-generation of reference
set from large quantity of small files, search shortcuts, etc.  Indexes are
now stored compressed on disk using lz4.

More additions to --verbose as progress indicators.

Overhaul of sdhash-cli options, matching sdhash as closely as possible and much 
more flexible.  Also removing last traces of posix from sdhash-cli for inclusion 
into windows port.
--------
New in version 3.0

sdbf_set indexing method, to be generated with --index while hashing
and searched with --search-index while hashing.

temp-bugfix: some speedup for very small reference data
bugfix: segmenting was seeing too-large final segment on windows
--------
New in version 2.3

Web interface using python/jquery -- talks to thrift
--validate option
tons of --verbose debugging output
New thrift-server sources directory
bugfix: for invalid sdbf file reading no longer crashes
bugfix: windows recursive search filesystem permissions crash fixed

--------
New in version 2.2

Very beta native win32/win64 port
Threads updated to use boost::thread instead of pthreads directly
Program options updated to support all command line options in sdhash.cfg
Filesystem access now primarily c++/boost not posix
Thrift client/server now supports asynchronous mode comparisions
output file option -o --output 

--------

New in version 2.1

recursive directory hashing with --deep -r option
sdhash supports input from stdin
argument --segment-size -z to customize read-in segment size
argument --name -n to rename stdin hashes
new client/server programs using Thrift
api updates to use const and clearer naming conventions
api updated with formal 'set' class
sdhash auto-switching to block mode when files are >16MB
new base64 encoding library

--------


New in version 2.0

gnu-style long-options for standalone and client programs
configuration file capability for standalone sdhash sdhash.cfg
fixed: sdhash buffers large files instead of memory mapping the entire thing
sdhash server program sdhashd, and sdhash-cli client program.
Client and server should be considered beta software.
sdhash-dd program removed and changed to -b option for block size
sdbf version upgraded to add 'original size' into the header information
"view" option to create subsets of sets matching an expression
added: Boost libraries for regular expressions
fixed: import as file list now can use DD mode
fixed: sampling now non-destructive
swig-python bindings now exist with a sample program
Now large block hashes are "streamed" in chunks instead of as one giant hash.

---------

New in version 1.8

-i file.txt option to generate hashes from a file list. 
conversion to C++ 
beta libsdbf API 
fixed: off-by-one error on last line of query file

---------

New in version 1.7

-c query.sdbf target.sdbf option
-s sample option
Support for cygwin
manpage for sdhash
make install supported
fixed: spaces in filenames now readable on input sdbf

---------


sdhash-server/src/sdhash-cli.cc

/**
 * sdhash-cli: Client for hashing server
 * author: Candice Quates
 */

#include <protocol/TBinaryProtocol.h>
#include <transport/TSocket.h>
#include <transport/TTransportUtils.h>

#include "boost/program_options.hpp"
#include "boost/lexical_cast.hpp"
#include <boost/filesystem.hpp>
#include <fstream>
#include <iostream>

#include "sdhashsrv.h"

using namespace std;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

using namespace sdhash;

using namespace boost;

namespace po = boost::program_options;
namespace fs = boost::filesystem;


#include "sdhash-cli.h"

// Global parameters
sdhash_parameters_t sdbf_sys = {
    (char*)HOST,        // hostname
    PORT,        // port#
    0,                // dd block size off initially
    1,               // output_threshold
    FLAG_OFF,        // warnings
    0,              // sample size off
    0,        // default no options set
    0,        // default no options2 set
    -1,        // result ID off
    (char*)"default"        // default username
};

int main( int argc, char **argv) {
    uint32_t  i, j, k, file_cnt;
    int set1;
    char tempstring[]="/tmp/default-XXXXXX";
    boost::filesystem::path current_dir=boost::filesystem::initial_path();
    int hashsetID = -1;
    vector<string> inputlist;
    string input_name;
    string user_name;
    string hostname;
    string config_file;
    string segment_size;
    int32_t resultID;
    po::variables_map vm;
    po::options_description config("Configuration");

    try {
        // Declare a group of options that will be 
        // allowed both on command line and in
        // config file
        config.add_options()
                ("config-file,C", po::value<string>(&config_file)->default_value("sdhash-cli.cfg"), "")
               // ("hash-list,f",po::value<std::string>(&listingfile),"generate SDBFs from list of filenames")
               // ("deep,r", "generate SDBFs from directories and files")
                ("host,H",po::value<std::string>(&hostname)->default_value("localhost"),"server to connect to")
                ("port,P",po::value<uint32_t>(&sdbf_sys.port)->default_value(9090),"server port")
                ("list,l","list all resident sets of SDBFs")
                ("compare,c","compare all pairs in SDBF file, or compare two SDBF files to each other")
                ("import,i","load SDBFs from file")
                ("export,e","export SDBFs to file <N> <file>.")
                ("results,r",po::value<int32_t>(&resultID),"retrieve result by ID")
                ("results-list","retrieve results list")
                ("type",po::value<std::string>(&user_name),"result type, default|indexing|web")
                ("show-set-data,d",po::value<int32_t>(&resultID),"retrieve result by ID")
                ("show-set-names,n",po::value<int32_t>(&resultID),"retrieve result by ID")
                ("threshold,t",po::value<int32_t>(&sdbf_sys.output_threshold)->default_value(1),"only show results >=threshold")
                ("block-size,b",po::value<uint32_t>(&sdbf_sys.dd_block_size),"hashes input files in nKB blocks")
                ("sample-size,s",po::value<uint32_t>(&sdbf_sys.sample_size)->default_value(0),"sample N filters for comparisons")
                ("name",po::value<std::string>(&input_name),"declare set name")
                ("index","generate indexes while hashing")
                ("index-search","match indexes at set level")
                ("verbose","turn on warnings")
                ("version","show version info")
                ("help,h","produce help message")
            ;
        // options that do not need to be in help message
        po::options_description hidden("Hidden");
        hidden.add_options()
            ("input-files", po::value<vector<std::string> >(&inputlist), "input file")
            ;

        po::options_description cmdline_options;
        cmdline_options.add(config).add(hidden);

        po::options_description config_file_options;
        config_file_options.add(config);

        // setup for list of files on command line
        po::positional_options_description p;
        p.add("input-files", -1);

        store(po::command_line_parser(argc, argv).
              options(cmdline_options).positional(p).run(), vm);
        notify(vm);

        std::ifstream ifs(config_file.c_str());
        if (vm.count("help")) {
            std::cout << VERSION_INFO << std::endl;
            std::cout << "Usage: sdhash-cli <options> <source files>|<hash files>"<< std::endl;
            std::cout << config << std::endl;
            return 0;
        }

        if (vm.count("version")) {
            std::cout << VERSION_INFO << std::endl;
            std::cout << "       http://sdhash.org, license Apache v2.0" << std::endl;
            return 0;
        }
        if (vm.count("warnings")) {
            sdbf_sys.warnings = 1;
        }
    if (vm.count("host")) {
        sdbf_sys.hostname=(char*)hostname.c_str();
    }
    if (vm.count("import"))
        sdbf_sys.options|=MODE_IMPORT;
    if (vm.count("export"))
        sdbf_sys.options|=MODE_EXPORT;
    if (vm.count("list"))
        sdbf_sys.options|=MODE_LIST;
    if (vm.count("show-set-names"))
        sdbf_sys.options|=MODE_DISP;
    if (vm.count("show-set-data"))
        sdbf_sys.options|=MODE_CONTENT;
    if (vm.count("compare"))
        sdbf_sys.options|=MODE_COMP;
    if (vm.count("results")) {
        sdbf_sys.options|=MODE_RESULT;
        sdbf_sys.resultid=resultID;
    }
    if (vm.count("results-list")) {
        sdbf_sys.options|=MODE_RESULT;
            if (vm.count("type"))
                sdbf_sys.username=(char*)user_name.c_str();
        sdbf_sys.resultid=0;
    }
    if (vm.count("type")) {
        sdbf_sys.username=(char*)user_name.c_str();
    }
    if (vm.count("index-search")) {
        sdbf_sys.options|=MODE_HASH;
        if (sdbf_sys.dd_block_size == 0);
            sdbf_sys.dd_block_size=16;
    }

    } catch (std::exception &e) {
        std::cout << e.what() << "\n";
        return 0;
    }

    int file_start=1;
    boost::shared_ptr<TTransport> socket(new TSocket(sdbf_sys.hostname, sdbf_sys.port));
    boost::shared_ptr<TTransport> transport(new TBufferedTransport(socket));
    boost::shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
    sdhashsrvClient client(protocol);
    if (sdbf_sys.options == 0 && vm.count("input-files")) {
        sdbf_sys.options|=MODE_HASH;
        if (vm.count("index-search")) {
            input_name = "searching"; 
        } else if (!vm.count("name")) {
            cerr << "SDHASH: hashset must be named with --name " << std::endl;
            return -1;
        }    
    }
    if (sdbf_sys.options == 0) {
        std::cout << VERSION_INFO << std::endl;
        std::cout << "Usage: sdhash-cli <options> <source files>|<hash files>"<< std::endl;
        std::cout << config << std::endl;
    return 0;
    }
    try {
        transport->open();
        client.ping();
        int i,fd;
        FILE *workfile;
        std::string result;
        uint32_t resultvalue = -2;
        switch (sdbf_sys.options) {
            case MODE_LIST:
                client.setList(result,false);
                break;
            case MODE_DISP:
                client.displaySet(result,resultID);
                break;
            case MODE_CONTENT:
                client.displayContents(result,resultID);
                break;
            case MODE_COMP:
                if (inputlist.size()==1) {
                    int32_t newid=client.createResultID(sdbf_sys.username);
                    int32_t set1id=boost::lexical_cast<int32_t>(inputlist[0]);
                    client.compareAll(set1id,sdbf_sys.output_threshold,newid);
                    std::cout << "Request name " << sdbf_sys.username <<", id: "<< newid << std::endl;
                } else if (inputlist.size()==2) {
                    int32_t newid=client.createResultID(sdbf_sys.username);
                    int32_t set1id=boost::lexical_cast<int32_t>(inputlist[0]);
                    int32_t set2id=boost::lexical_cast<int32_t>(inputlist[1]);
                    client.compareTwo(set1id,set2id,sdbf_sys.output_threshold,sdbf_sys.sample_size,newid);
                    std::cout << "Request name " << sdbf_sys.username <<", id: "<< newid << std::endl;
                }
                break;
            case MODE_IMPORT:
                if (inputlist.size()==1) {
                            resultvalue=client.loadSet(inputlist[0].c_str(),-1);
                }
                break;
            case MODE_INFILE:
                hashsetID = client.createHashsetID();
                //    client.hashList(argv[file_start],sdbf_sys.dd_block_size,hashsetID,(sdbf_sys.options2 & MODE_INDEX) );
                result="Hashing in progress for "+(string)argv[file_start];
                break;
            case MODE_HASH:
                // change to hash_string to make windows compat
                hashsetID = client.createHashsetID();
                if (vm.count("index")) {
                    client.hashString(input_name,inputlist,sdbf_sys.dd_block_size,hashsetID,-1);
                    result="Hashing in progress for "+input_name+".sdbf";
                }
                else if (vm.count("index-search")) {
                    int resultID= client.createResultID("indexing");
cout << sdbf_sys.dd_block_size;
                    client.hashString(input_name,inputlist,sdbf_sys.dd_block_size,resultID,3);
                    result="Searching in progress, result id: "+boost::lexical_cast<string>(resultID);
                }
                else { // plain 
                    client.hashString(input_name,inputlist,sdbf_sys.dd_block_size,hashsetID,0);
                    result="Hashing in progress for "+input_name+".sdbf";
                }
                break;
            case MODE_EXPORT:
                if (file_cnt == 2) {
                    if (isdigit(argv[file_start][0])) {
                        resultvalue=client.saveSet(atoi(argv[file_start]),argv[file_start+1]);
                        if (resultvalue==0) {
                            result = "Set saved";
                         }    
                    }
                }
                break;
            case MODE_RESULT: 
                if (sdbf_sys.resultid == 0) {
                    client.displayResultsList(result,sdbf_sys.username,false);
                } else {
                    client.displayResult(result,sdbf_sys.resultid);
                }
                break;
        }
        transport->close();
        if (result.length()==0) {
             if (resultvalue==-1) {
                std::cout << "ERROR: Command failed" << std::endl;
            } else if (resultvalue==-2) {
                //std::cout << "No results found" << std::endl;
            } else{
                std::cout << "Set "<< resultvalue << " created" << std::endl;
            }
        } else {
            std::cout << result << std::endl;
        }
    } catch (TException &tx) {
        printf("Couldn't connect: %s\n", tx.what());
    }
    return 0;
}
